src/kerberos/client/mod.rs

```diff
@@ -1,2 +1,435 @@
+                }
+            };
+            client.realm = Some(realm.clone());
+
+            let options = GenerateAsReqOptions {
+                realm: &realm,
+                username: &username,
+                cname_type,
+                snames: &[TGT_SERVICE_NAME, &realm],
+                // 4 = size of u32
+                nonce: &OsRng.gen::<[u8; 4]>(),
+                hostname: &unwrap_hostname(client.config.client_computer_name.as_deref())?,
+                context_requirements: builder.context_requirements,
+            };
+            let kdc_req_body = generate_as_req_kdc_body(&options)?;
+
+            let pa_data_options = match credentials {
+                CredentialsBuffers::AuthIdentity(auth_identity) => {
+                    let domain = utf16_bytes_to_utf8_string(&auth_identity.domain);
+                    let salt = format!("{}{}", domain, username);
+
+                    AsReqPaDataOptions::AuthIdentity(GenerateAsPaDataOptions {
+                        password: &password,
+                        salt: salt.as_bytes().to_vec(),
+                        enc_params: client.encryption_params.clone(),
+                        with_pre_auth: false,
+                    })
+                }
+                CredentialsBuffers::SmartCard(smart_card) => {
+                    let private_key_pem = utf16_bytes_to_utf8_string(
+                        smart_card
+                            .private_key_pem
+                            .as_ref()
+                            .ok_or_else(|| Error::new(ErrorKind::InternalError, "scard private key is missing"))?,
+                    );
+                    client.dh_parameters = Some(generate_client_dh_parameters(&mut OsRng)?);
+
+                    AsReqPaDataOptions::SmartCard(Box::new(pk_init::GenerateAsPaDataOptions {
+                        p2p_cert: picky_asn1_der::from_bytes(&smart_card.certificate)?,
+                        kdc_req_body: &kdc_req_body,
+                        dh_parameters: client.dh_parameters.clone().unwrap(),
+                        sign_data: Box::new(move |data_to_sign| {
+                            let mut sha1 = Sha1::new();
+                            sha1.update(data_to_sign);
+                            let hash = sha1.finalize().to_vec();
+                            let private_key = PrivateKey::from_pem_str(&private_key_pem)?;
+                            let rsa_private_key = RsaPrivateKey::try_from(&private_key)?;
+                            Ok(rsa_private_key.sign(Pkcs1v15Sign::new::<Sha1>(), &hash)?)
+                        }),
+                        with_pre_auth: false,
+                        authenticator_nonce: OsRng.gen::<[u8; 4]>(),
+                    }))
+                }
+            };
+
+            let as_rep = as_exchange(client, yield_point, &kdc_req_body, pa_data_options).await?;
+
+            info!("AS exchange finished successfully.");
```

Comment:
issue: Please avoid using INFO-level logs in libraries, as it tends to pollute usual logs of applications such as IronRDP and Devolutions Gateway. Let’s use DEBUG and TRACE only in the libraries. Feel free to use INFO in the binaries such as the `ffi` crate and other examples.